internal/pkg/plugin/registry.go

```diff
@@ -18,62 +18,85 @@
 package plugin
 
 import (
+	"errors"
 	"fmt"
 	"reflect"
 	"sync"
 )
 
-// All plugins is wrote in ./plugins dir. The plugin type would be as the next level dirs,
-// such as collector, client, or queue. And the 3rd level is the plugin name, that is also
-// used as key in pluginRegistry.
-
-// reg is the global plugin registry
+// the global plugin registry
 var (
```

Comment:
Is this really required to be global? Take a look at Java version's `ModuleManager`. Once you accept the global singleton, the tests are sensitive about clearing this register table after every time, and you don't have a chance to run tests concurrently(even I don't know whether this is an option in go).

Reply:
It should be global because the registered way is func init()

Reply:
This leads to another question, why needs the `package init()` mechanism? All the plugins should be registered due to the configuration files or parameters, aren't they? Otherwise, you will cost the memory for nothing.

Reply:
Golang app has nothing like SPI in Java, a hardcode `if/else` section could take care of the register easily. Better to read, and easy to extend, no extra memory cost to maintain this register table.

Reply:
the table only registers the reflect.type, I don't think this would cost resources for future extensions.

Reply:
A similar SPI mechanism is implemented, I don't think hard code to register plugins is a good way, Similar patterns exist in many places, such as mosn, beats, and etc... . In the original version, I also experimented with the golang dynamic link(.so mechanism), but it was difficult to use

Reply:
there is a demo to compute the size of (reflect.type), the size is only few Bytes, https://stackoverflow.com/questions/31338588/generic-function-to-get-size-of-any-structure-in-golang

Reply:
I know you are writing an SPI mechanism in the structure. Personally, I have concerns about that. 
This thing makes the codes more complex than it needs to be, at the same time, w/o the dynamic link(I don't ask for implementing dynamic link) this mechanism provides no new feature.
This is the reason mostly.

Also, around the tech part
> In Go, the predefined init() function sets off a piece of code to run before any other part of your package. This code will execute as soon as the package is imported.

How could you sure the package is imported if the implementation in a random package? Or some packages the core doesn't know to access in the initializing stage?

___
In the basic idea, I don't want the go works like Java. @hanahmily @kezhenxu94 I hope to get your inputs.

Reply:
if a specific plugin want to register itself to the registry, it must invoke the register method in its init func. That purpose the specific plugin must import the registry package. In the golang, the import package init would be first invoked. The reason why i want to use init func is it is hard to control the used plugins by manual registration.

Reply:
Satellite supports multi plugins ，so we support a registration mechanism to control this plugins. So there are 2 ways to register the plugins, which are auto registration and manual registration. Auto registration is relay on the init func in golang, that makes the codes more cleaner. However, it also makes the codes more complex. Another ways is manually registration, that would add few go files to control the plugins registration. This way would easier to read source code, but would add some files to control the registration. Which one should we choose？

Reply:
Note, auto-registration would cost clear memory, even limited, our SLO is 20M for the whole satellite deployment, consider this in the decision.

Reply:
Please edit this comment to show your preference. @apache/skywalking-committers, especially reviewers, @surlymo @hanahmily @kezhenxu94 @dmsolr @zhangkewei @gxthrj @nic-chen @moonming 

- +1 for Auto-registration. 
- +1 for manual registration in go source codes. @wu-sheng

Reply:
> This leads to another question, why needs the `package init()` mechanism? All the plugins should be registered due to the configuration files or parameters, aren't they? Otherwise, you will cost the memory for nothing.

+1

Reply:
ok, I would change to the manual way.

Reply:
+1 for manual registration in go source codes

i think either is ok. auto-registration like spring annotation declaration in java, manual registration like spring xml definetion. i prefer the latter.